src/knot_thing_main.c

```diff
@@ -6,196 +6,160 @@
-		isensor[index] = 0;
-
-	for (index = 0; index < KORE_FLOAT_SENSORS; index++)
-		fsensor[index] = 0;
-
-	for (index = 0; index < KORE_BOOL_SENSORS; index++)
-		bsensor[index] = 0;
-
-	for (index = 0; index < KORE_RAW_SENSORS; index++)
-		rsensor[index] = 0;
-
-	return 0;
+	return -1;
 }
 
-void kore_exit(void)
+uint8_t item_is_unregistered(uint8_t sensor_id)
 {
-
+	return (!(data_items[sensor_id].event_flags & KNOT_EVT_FLAG_UNREGISTERED));
 }
 
-int8_t kore_run(void)
+int8_t knot_thing_init(void)
 {
-	int32_t hash_value, value_int = 0, value_dec = 0, multiplier = 0;
-	int8_t index, value_bool;
-
-	/* TODO: Monitor events from network and sensors */
-
-	/*
-	 * For all registered integer sensors: verify if value
-	 * from sensor has changed based on the last value read.
-	 */
-
-	/* For integer sensors: value changed? */
-	for (index = 0; index < isensor_num; index++) {
-
-		if (!isensor[index]->read)
-			continue;
-
-		if (isensor[index]->read(&value_int, &multiplier) < 0)
-			continue;
-
-		hash_value = hash(value_int, multiplier, 0);
-
-		if (hash_value == ihash[index])
-			continue;
-
-		ihash[index] = hash_value;
-	}
-
-	/* For float sensors: value changed? */
-	for (index = 0; index < fsensor_num; index++) {
-
-		if (!fsensor[index]->read)
-			continue;
-
-		if (fsensor[index]->read(&value_int, &multiplier,
-							&value_dec) < 0)
-			continue;
-
-		hash_value = hash(value_int, multiplier, value_dec);
-		if (hash_value == fhash[index])
-			continue;
-
-		fhash[index] = hash_value;
-	}
-
-	/* For boolean sensors: value changed? */
-	for (index = 0; index < bsensor_num; index++) {
-
-		if (!bsensor[index]->read)
-			continue;
-
-		if (bsensor[index]->read(&value_bool) < 0)
-			continue;
-
-		if (value_bool = bhash[index])
-			continue;
-
-		bhash[index] = value_bool;
-	}
-
-	/* For raw sensors: value changed? */
-	for (index = 0; index < rsensor_num; index++) {
-
-		if (!rsensor[index]->read)
-			continue;
-		/*
-		 * TODO: Postpone raw data change verification. It
-		 * requires a fast hash function for binary data.
-		 */
-	}
-
+	reset_data_items();
 	return 0;
 }
 
-int8_t kore_sensor_register_integer(struct sensor_integer *sensor)
+void knot_thing_exit(void)
 {
-	 /*
-	  * If KORE_INTEGER_SENSORS (0) or isensor_num reached
-	  * the user defined amount: return no memory.
-	  */
-	if (isensor_num == KORE_INTEGER_SENSORS)
-		return -ENOMEM;
-
-	isensor[isensor_num] = sensor;
-	isensor_num++;
 
-	return 0;
 }
 
-int8_t kore_sensor_register_float(struct sensor_float *sensor)
+int8_t knot_thing_register_data_item(uint8_t sensor_id, const char *name, uint16_t type_id,
+	uint8_t value_type, uint8_t unit, knot_data_functions * func)
 {
-	 /*
-	  * If KORE_FLOAT_SENSORS (0) or fsensor_num reached
-	  * the user defined amount: return no memory.
-	  */
-	if (fsensor_num == KORE_FLOAT_SENSORS)
-		return -ENOMEM;
-
-	fsensor[fsensor_num] = sensor;
-	fsensor_num++;
+	if (sensor_id < KNOT_THING_DATA_MAX)
```

Comment:
nitpick (it should not block this code to be pushed upstream): invert the logic helps to reduce indentation.